autoversion_lib/__init__.py

```diff
@@ -156,7 +156,10 @@ def commit2timestamp(self, commit: str) -> int:
 
 class Autoversion:
     def __init__(self, path: str, precache: bool=False):
-        git = Git(path)
+        try:
+            git = Git(path)
+        except GitError as e:
+            raise AutoverError(e)
```

Comment:
What type is `AutoverError.args[0]` meant  to be? It appears from all other uses it is expected to be a string. So perhaps `raise AutoverError("an error occured initializing gitwrapper") from e`

Reply:
(Part of this discussion was had offline - apologies if there is missing context)

To explicitly respond to your question, TBH I have no idea what `AutoverError.args[0]` is meant to be?! I've just spent a while reading about python exceptions and from what I understand, (at least by convention) `Exception.args[0]` is always string?!

So if I wanted to leave the GitError text as is, are you suggesting that I should pass it as a string? I.e. like this:

```
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError(f'{e}') from e
```
FWIW, that returns this:
```
Traceback (most recent call last):
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 160, in __init__
    git = Git(path)
  File "/usr/lib/python3/dist-packages/gitwrapper/__init__.py", line 171, in __init__
    raise GitError("Not a git repository `%s'" % self.path)
gitwrapper.GitError: Not a git repository `/turnkey/public/autoversion/autoversion_lib'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/bin/autoversion", line 74, in <module>
    main()
  File "/usr/bin/autoversion", line 60, in main
    auto_ver = autoversion.Autoversion(os.getcwd(), precache=args.reverse)
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 162, in __init__
    raise AutoverError(f'{e}') from e
autoversion_lib.AutoverError: Not a git repository `/turnkey/public/autoversion/autoversion_lib'
```
Or do you mean something else?

TBH, I don't really understand what you meant (in our offline discussion). If I don't capture the exception in the CLI (via the `try...except`) then I get this instead:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 160, in __init__
    git = Git(path)
  File "/usr/lib/python3/dist-packages/gitwrapper/__init__.py", line 171, in __init__
    raise GitError("Not a git repository `%s'" % self.path)
gitwrapper.GitError: Not a git repository `/turnkey/public/autoversion/autoversion_lib'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/autoversion", line 74, in <module>
    main()
  File "/usr/bin/autoversion", line 60, in main
    auto_ver = autoversion.Autoversion(os.getcwd(), precache=args.reverse)
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 162, in __init__
    raise AutoverError(e)
autoversion_lib.AutoverError: Not a git repository `/turnkey/public/autoversion/autoversion_lib'
```
 Which seems more-or-less the same as what you suggest?!

I do recall you mentioning the `from` thing, so perhaps it should be like this:
```
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError from e
```
If I do that though, the stacktrace I get is this:
```
Traceback (most recent call last):
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 160, in __init__
    git = Git(path)
  File "/usr/lib/python3/dist-packages/gitwrapper/__init__.py", line 171, in __init__
    raise GitError("Not a git repository `%s'" % self.path)
gitwrapper.GitError: Not a git repository `/turnkey/public/autoversion/autoversion_lib'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/bin/autoversion", line 74, in <module>
    main()
  File "/usr/bin/autoversion", line 60, in main
    auto_ver = autoversion.Autoversion(os.getcwd(), precache=args.reverse)
  File "/usr/lib/python3.9/dist-packages/autoversion_lib/__init__.py", line 162, in __init__
    raise AutoverError from e
autoversion_lib.AutoverError
```
Which also seems ok, although the AutoverError bit isn't quite as verbose.

What am I missing?

Reply:
Ok, so a bit more reading and thinking about this, here's what I am proposing:
```
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError(str(e)) from e
```
Although I'm not sure if the `from e` makes sense in this context?

Reply:
So just to clear things up, this isn't currently causing a bug of any kind. It's just that by convention, both for exceptions generally and in our uses of `AutoverError`, `args[0]` is a string. But `GitError` isn't a string. 

Your solution works:

```python
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError(str(e)) from e
```

in most cases, this will be the same as:

```python
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError(e.args[0]) from e
```

However I'd be tempted to add in some additional context, as when printed, this exception will look identical to the original GitError, which might make debugging more difficult.

So I'd probably do something like this, I'll leave it up to you whether or not it's worth it though.

```python
        try:
            git = Git(path)
        except GitError as e:
            raise AutoverError(f"failed to initialize gitwrapper: {e}") from e
```

Reply:
Ok great, thanks